### PR TITLE
Applying expansion for material composition changes

### PR DESCRIPTION
**Pull request by mgjarrett**
## What is the change?

When a composition is updated, check to see if this change affects the thermal expansion properties of the material. If it does, clear the cache to allow volumes to be recomputed and apply a scaling factor to the number densities to conserve mass.

## Why is the change being made?

Material properties can be dependent upon the composition of a material. When the composition of a material changes from depletion (or anything else), the thermal expansion properties can change, which causes a change in volume. This change looks like thermal expansion, but caused by a change in composition rather than a change in temperature. To accurately conserve mass throughout a simulation, it is beneficial to check for this type of expansion any time a composition changes and adjust the number densities accordingly.

---

## Checklist

- [x] This PR has only [one purpose or idea](https://terrapower.github.io/armi/developer/tooling.html#one-idea-one-pr).
- [x] [Tests](https://terrapower.github.io/armi/developer/tooling.html#test-it) have been added/updated to verify any new/changed code.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/developer/tooling.html#add-release-notes) have been updated if necessary.
- [x] The [documentation](https://terrapower.github.io/armi/developer/tooling.html#document-it) is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `pyproject.toml`.

### CODE REVIEW

File: armi/reactor/components/component.py [hunk 1]
```diff
@@ -369,11 +369,21 @@ def applyMaterialMassFracsToNumberDensities(self):
...
         density = self.material.getProperty("pseudoDensity", Tc=self.temperatureInC)
-
         self.p.numberDensities = densityTools.getNDensFromMasses(
             density, self.material.massFrac
         )
 
+        # Sometimes the material thermal expansion depends on its parent's composition (e.g. Pu frac)
+        # so setting number densities can sometimes change thermal expansion behavior.
+        # Call again so that the material has access to its parent's comp when providing the reference
+        # initial density.
```
Comment (john-science) [inline armi/reactor/components/component.py:379]:
```suggestion
        # Sometimes material thermal expansion depends on its parent's composition (e.g. Pu frac) so
        # setting number densities can sometimes change thermal expansion behavior. Call again so
        # the material has access to its parent's comp when providing the reference initial density.
```

File: armi/reactor/components/component.py [hunk 2]
```diff
@@ -738,26 +757,62 @@ def setNumberDensities(self, numberDensities):
...
+        -----
+        Sometimes volume/dimensions can change due to the number density change when the material thermal
+        expansion depends on the component's composition (e.g. its plutonium fraction). In this case, changing the
+        density will implicitly change the area/volume. Since it is very difficult to predict the new dimensions
+        ahead of time, and perturbation/depletion calculations are almost exclusively done assuming constant volume,
+        the densities sent are automatically adjusted to conserve mass with the original dimensions. That is, the
+        component's densities are not exactly as passed, but whatever they would need to be to preserve volume
+        integrated number densities (moles) from the pre-perturbed component's volume/dimensions.
+        Note this has no effect if the material thermal expansion has no dependence on component composition.
+        If this is not desired, `self.p.numberDensities` can be set directly.
```
Comment (john-science) [inline armi/reactor/components/component.py:785]:
```suggestion
        Sometimes volume/dimensions change due to number density change when the material thermal
        expansion depends on the component's composition (e.g. its plutonium fraction). In this
        case, changing the density will implicitly change the area/volume. Since it is difficult to
        predict the new dimensions, and perturbation/depletion calculations almost exclusively
        assume constant volume, the densities sent are automatically adjusted to conserve mass with
        the original dimensions. That is, the component's densities are not exactly as passed, but
        whatever they would need to be to preserve volume integrated number densities (moles) from
        the pre-perturbed component's volume/dimensions.

        This has no effect if the material thermal expansion has no dependence on component
        composition. If this is not desired, `self.p.numberDensities` can be set directly.
```

File: armi/reactor/components/component.py [hunk 3]
```diff
@@ -738,26 +757,62 @@ def setNumberDensities(self, numberDensities):
...
-        # since we're updating the object the param points to but not the param itself, we have to inform
+
+        # check if thermal expansion changed
+        dLLnew = self.material.linearExpansionPercent(Tc=self.temperatureInC) / 100.0
+        if dLLprev != dLLnew and materialExpansion:
+            # the thermal expansion changed so the volume change is happening at same time as
+            # density change was requested. Attempt to make mass consistent with old dims
+            # (since the density change was for the old volume and otherwise mass wouldn't be conserved)
+
+            # enable recalculation of volume, otherwise it uses stored.
```
Comment (john-science) [inline armi/reactor/components/component.py:810]:
```suggestion
            # the thermal expansion changed so the volume change is happening at same time as
            # density change was requested. Attempt to make mass consistent with old dims (since the
            # density change was for the old volume and otherwise mass wouldn't be conserved).

            # enable recalculation of volume, otherwise it uses stored
```

**Comment by john-science** [comment]
@mgjarrett Just curious why this is still draft.

File: armi/reactor/components/component.py [hunk 4]
```diff
@@ -651,6 +660,20 @@ def setTemperature(self, temperatureInC):
         self.changeNDensByFactor(f)
         self.clearLinkedCache()
 
+    def changeNDensByFactor(self, factor):
```
Comment (john-science) [inline armi/reactor/components/component.py:663]: Can we get a unit test to cover this new method and its edge cases?

File: armi/reactor/components/component.py [hunk 5]
```diff
@@ -734,22 +752,71 @@ def setNumberDensities(self, numberDensities):
...
+            area = self.getArea()
+
+        # change the densities
+        if wipe:
+            self.p.numberDensities = {}  # clear things not passed
         self.p.numberDensities.update(numberDensities)
-        # since we're updating the object the param points to but not the param itself, we have to inform
-        # the param system to flag it as modified so it properly syncs during ``syncMpiState``.
-        self.p.assigned = parameters.SINCE_ANYTHING
-        self.p.paramDefs["numberDensities"].assigned = parameters.SINCE_ANYTHING
```
Comment (john-science) [inline armi/reactor/components/component.py:752]: We can just remove this without hurting parallel runs? That's surprising.

Comment (mgjarrett) [reply armi/reactor/components/component.py:752]: Nope, this caused a bug. reverting now...

Comment (john-science) [reply armi/reactor/components/component.py]: Ah, okay, that makes more sense.

Thanks!

File: armi/reactor/components/component.py [hunk 6]
```diff
@@ -734,23 +738,93 @@ def setNumberDensities(self, numberDensities):
...
+            else:
+                factor = area / self.getArea()
+            self.changeNDensByFactor(factor)
+
         # since we're updating the object the param points to but not the param itself, we have to inform
         # the param system to flag it as modified so it properly syncs during ``syncMpiState``.
         self.p.assigned = parameters.SINCE_ANYTHING
         self.p.paramDefs["numberDensities"].assigned = parameters.SINCE_ANYTHING
 
+    def changeNDensByFactor(self, factor):
```
Comment (john-science) [inline armi/reactor/components/component.py:812]: Can we get a unit test for the new methods we are adding?

Thanks!

Comment (drewj-tp) [reply armi/reactor/components/component.py:825]: added

File: armi/reactor/components/component.py [hunk 7]
```diff
@@ -747,23 +751,93 @@ def setNumberDensities(self, numberDensities):
...
         self.p.numberDensities.update(numberDensities)
+
+        # check if thermal expansion changed
+        dLLnew = self.material.linearExpansionPercent(Tc=self.temperatureInC) / 100.0
+        if dLLprev != dLLnew and materialExpansion:
+            # the thermal expansion changed so the volume change is happening at same time as
+            # density change was requested. Attempt to make mass consistent with old dims (since the
+            # density change was for the old volume and otherwise mass wouldn't be conserved).
+
+            self.clearLinkedCache()  # enable recalculation of volume, otherwise it uses cached
```
Comment (john-science) [inline armi/reactor/components/component.py:813]: I am curious how this effects our performance. It is not immediately obvious to me what will happen. 

Hmmm

File: armi/reactor/tests/test_components.py [hunk 8]
```diff
@@ -49,6 +50,27 @@
     materials,
 )
 from armi.testing import loadTestReactor
+from armi.utils.units import getTc
+
+
+class CompositionDependentExpander(materials.Material):
```
Comment (john-science) [inline armi/reactor/tests/test_components.py:56]: I apologize for myself.

But can we rename this something like `MockCompositionDependentExpander`?

I know that seems like a pedantic ask. But, no joke, we have had people import things from out tests classes for non-test uses in the past. And... #sigh. That if words like "mock" or "test" or "dummy" are included, I feel like we've done all we can.

Sorry.

**Review by john-science — APPROVED** [review]
I would like that test class rename. But I guess if you all are in a dead rush, I can open up a follow up PR about it.
